Packages/com.llealloo.audiolink/Editor/Scripts/AudioLinkShaderCompatabilityUtility.cs

```diff
@@ -0,0 +1,125 @@
+            + "\n" + "\n" +
+            "(You can always run this utility manually via the " + MenuItemPath + " menu command)";
+
+        private const string DialogTitle = "Upgrade AudioLink compatible shaders";
+        private const string DialogOkButton = "I Made a Backup, Go Ahead!";
+        private const string DialogCancelButton = "No Thanks";
+
+
+        [MenuItem(MenuItemPath)]
+        public static void UpgradeShaders()
+        {
+            if (EditorUtility.DisplayDialog(DialogTitle, DialogText, DialogOkButton, DialogCancelButton))
+            {
+                UpgradeShaderFiles();
+                UpgradeCgincFiles();
+                AssetDatabase.Refresh();
+            }
+        }
+
+        private static void UpgradeShaderFiles()
+        {
+            ShaderInfo[] shaders = ShaderUtil.GetAllShaderInfo();
+
+            foreach (ShaderInfo shaderinfo in shaders)
+            {
+                Shader shader = Shader.Find(shaderinfo.name);
+                string path = AssetDatabase.GetAssetPath(shader);
+                // we want to avoid built-in shaders so we check the Path
+                if (path.StartsWith("Assets") || path.StartsWith("Packages"))
+                {
+                    ReplaceInFile(path);
+                }
+            }
+        }
+
+        private static void UpgradeCgincFiles()
+        {
+            List<string> cgincs = FindCgincFiles(Application.dataPath);
+            cgincs.AddRange(FindCgincFiles(Path.GetDirectoryName(Application.dataPath) + "/Packages"));
+
+            foreach (string cginc in cgincs)
+            {
+                ReplaceInFile(cginc);
+            }
+        }
+
+        private static List<string> FindCgincFiles(string path)
+        {
+            List<string> cgincList = new List<string>();
+
+            string[] files = Directory.GetFiles(path);
+            foreach (string file in files)
+            {
+                if (file.EndsWith(".cginc") || file.EndsWith(".hlsl"))
+                {
+                    cgincList.Add(file);
+                }
+            }
+
+            string[] folders = Directory.GetDirectories(path);
+            foreach (string folder in folders)
+            {
+                List<string> cgincs = FindCgincFiles(folder);
+                cgincList.AddRange(cgincs);
+            }
+
+            return cgincList;
+        }
+
+        private static void ReplaceInFile(string path)
+        {
+            string[] shaderSource = File.ReadAllLines(path);
+            bool shouldWrite = false;
+            for (int i = 0; i < shaderSource.Length; i++)
+            {
+                string line = shaderSource[i];
+                if (line.Contains(OldAbsolutePath))
+                {
+                    line = line.Replace(OldAbsolutePath, NewAbsolutePath);
+                    while (line.Contains("/Packages"))
```

Comment:
What is this

Reply:
therefore we have to remove any prefixed /s

Reply:
Is that really what this code is doing, though? The replace call would replace 'Packages' anywhere in the string, and Contains would check anywhere in the string. I'm assuming you got this right, I just wanna hear your thoughts.

Reply:
I've tested this thoroughly, maybe you can re-read it.
this is just to make sure we don't have `#include \Packages\` because we essentially do `line.Replace("Assets","Packages")` so if it has a suffixed '\' it's then illegal